File: 13주차/KM/17-letter-combinations-of-a-phone-number.java
import java.util.List;
import java.util.ArrayList;
class Solution {
    String[] letters = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    List<String> answer = new ArrayList<>();
    boolean[] visited = new boolean[4]; 
    public List<String> letterCombinations(String digits) {
        if(digits.length() == 0) return answer;
        Combinations("", digits);  
        return answer;
    }
    public void Combinations(String result, String digits){
        if(result.length() == digits.length()){
            answer.add(result);
        } else {
            for(int i = result.length(); i<digits.length(); i++){
                String letter = letters[digits.charAt(i)- '0' - 2];
                for(int j = 0; j<letter.length(); j++){
                    if(!visited[i]){
                        visited[i] = true;
                        Combinations(result+ letter.charAt(j), digits);
                        visited[i] = false;
                    }
                    
                }
            }
        }
    }
}

File: 13주차/DE/17. Letter Combinations of a Phone Number.js
/**
 * @param {string} digits
 * @return {string[]}
 */


 function letterCombinations(digits) {
    let num = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
    if(digits == ''){
        return []
    }
    
    let result = ['']
    
    /*
    while문 ver
    while(digits){
        let digit = num[Number(digits[0])].split("");
        let temp = []
        for(let i = 0; i < result.length; i++){
            for(let j = 0; j < digit.length; j++) {
                temp.push(result[i] + digit[j])
            }
        }
        result = temp;
        
        digits = digits.slice(1, digits.length)
    }
    */
    //for문 ver
    for(let idx = 0; idx < digits.length; idx++){
        let digit = num[Number(digits[idx])].split("");
        let temp = []
        for(let i = 0; i < result.length; i++){
            for(let j = 0; j < digit.length; j++) {
                temp.push(result[i] + digit[j])
            }
        }
        result = temp;
    }
    
    return result;
};